### DISCUSSION TITLE
Encoding issues with zos_job_submit

### DISCUSSION THREAD

**Discussion by ghost** [ansible-collections/ibm_zos_core#1003] [Q&A]
I'm having a coding problem when using the zos_job_submit module.
I have two tasks that perform practically the same thing: edit a unix file, copy it to a pds library as a member and then submit the job.
One of the tasks is executed normally.
The other returns the following error:

![image](https://github.com/ansible-collections/ibm_zos_core/assets/105386204/f2ea8ae3-4ab0-4adc-80f0-0cd566efe8dc)

I've already tried changing the encoding everywhere (when copying from unix to pds, when running the job, etc.). I used everything that made any sense (utf-8, iso8859-1, ibm1047).
I've already checked the pds profile and they are exactly the same (the job that works and the one that doesn't).
I checked in the job that doesn't work if there is any special character that could be causing a problem and I didn't find it either.

This is the job submited and thats ends ok:
[CICALOC.txt](https://github.com/ansible-collections/ibm_zos_core/files/12769038/CICALOC.txt)

This is the job that receive the error:
[CICEZAC.txt](https://github.com/ansible-collections/ibm_zos_core/files/12769042/CICEZAC.txt)

Does anyone have any ideas on how I can identify what is causing this?

**Comment by ddimatos** [comment]
Hi @fabiobncbb - I believe you have encountered this issue #677, my assumption is what is happening is that there is nothing wrong with the job such that you should be trying to change the encoding, the issue I believe is that the jobs output contains a non-printable UTF-8 char and when Python encounters a character it can not convert, it prints this error. According to this error its the start byte in the job log, try to view the job in HEX and then search a UTF-8 list of HEX values and likely what you will find is that there does not exist a corresponding value, hence the error.

I often see this happen when compile and linking is in the job such as your second posted job which is performing a LINKEDIT.

The fix will come in approximately at the end of Q1 2024.

Some other work arounds might be to try to use `zos_mvs_raw` (port the JCL to it) and have the SYSPRINT be written to a data set, I have not tried this, the other option is to invoke JCL with native USS or ZOAU options and avoid a response with the non-printable character.

**Comment by ghost** [comment]
Hi @ddimatos !
Thank you very much for your attention!

I believe this is exactly my problem!
I managed to get around the problem by capturing the job number that appears in the message with a regular expression and showing the syout only of the ddname "JESMSGLG", which does not contain any characters that cannot be decoded by UTF-8. This already meets the needs of my task!

![image](https://github.com/ansible-collections/ibm_zos_core/assets/105386204/98f91a08-1ff3-4a3f-8871-d55a025a0623)

![image](https://github.com/ansible-collections/ibm_zos_core/assets/105386204/e0cdf64b-11f5-4a03-949e-abd859a71c18)

**Reply by ddimatos** [reply]
@fabiobncbb - that is a good and reasonable work around (so long as users don't need the full job log returned). Thanks for sharing your work around, it could very well help others.
